src/boot/z_std_dma.c

```diff
@@ -0,0 +1,468 @@
+    }
+
+    sprintf(buff2, "%07X %08X %X %.50s", vrom, vram, size, (file != NULL) ? file : "???");
+    Fault_AddHungupAndCrashImpl(buff1, buff2);
+}
+
+/**
+ * Transfer `size` bytes from physical ROM address `vrom` to `vram`.
+ *
+ * This function is intended for internal use only, however it is possible to use this function externally in which
+ * case it behaves as a synchronous transfer, data is available as soon as this function returns.
+ *
+ * Transfers are divided into chunks based on the current value of `gDmaMgrDmaBuffSize` to avoid congestion of the PI
+ * so that higher priority transfers can still be carried out in a timely manner. The transfers are sent in a queue to
+ * the OS PI Manager which performs the transfer.
+ *
+ * The `vrom` address and the `size` are expected to be at least 0x2 aligned, while the destination `vram` should be 0x8
+ * aligned, otherwise a fatal error will be triggered.
+ *
+ * @return 0 if successful, -1 if the DMA could not be queued with the PI Manager.
+ */
+s32 DmaMgr_DmaRomToRam(RomOffset vrom, void* vram, size_t size) {
+    OSIoMesg ioMsg;
+    OSMesgQueue queue;
+    OSMesg msg[1];
+    s32 ret;
+    size_t buffSize = gDmaMgrDmaBuffSize;
+    DmaRequest sp3C;
+
+    if (((vrom << 31) != 0) || (((u32)vram << 29) != 0) || ((size << 31) != 0)) {
+        sp3C.vrom = vrom;
+        sp3C.vram = vram;
+        sp3C.size = size;
+        sp3C.filename = "percial_DMA";
+        sp3C.line = 0;
+        // "alignment error"
+        DmaMgr_Error(&sp3C, NULL, "ILLIGAL ALIGNMENT", "アライメント異常");
+    }
+
+    osInvalDCache(vram, size);
+    osCreateMesgQueue(&queue, msg, ARRAY_COUNT(msg));
+
+    if (buffSize > 0) {
+        while (size > buffSize) {
+            // The system avoids large DMAs as these would stall the PI for too long, potentially causing issues with
+            // audio. To allow audio to continue to DMA whenever it needs to, other DMAs are split into manageable
+            // chunks.
+
+            ioMsg.hdr.pri = 0;
+            ioMsg.hdr.retQueue = &queue;
+            ioMsg.devAddr = vrom;
+            ioMsg.dramAddr = vram;
+            ioMsg.size = buffSize;
+            B_800406B0_jp += buffSize;
+
+            //! FAKE
+            if (1) {}
+            if (1) {}
+
+            ret = osEPiStartDma(carthandle, &ioMsg, 0);
+            if (ret != 0) {
+                goto end;
+            }
+
+            osRecvMesg(&queue, NULL, OS_MESG_BLOCK);
+            size -= buffSize;
+            vrom += buffSize;
+            vram = (void*)((uintptr_t)vram + buffSize);
+        }
+    }
+
+    ioMsg.hdr.pri = 0;
+    ioMsg.hdr.retQueue = &queue;
+    ioMsg.devAddr = vrom;
+    ioMsg.dramAddr = vram;
+    ioMsg.size = size;
+    B_800406B0_jp += size;
+
+    ret = osEPiStartDma(carthandle, &ioMsg, 0);
+    if (ret == 0) {
+        osRecvMesg(&queue, NULL, OS_MESG_BLOCK);
+        osInvalDCache(vram, size);
+    }
```

Comment:
```suggestion
    if (ret != 0) {
        goto end;
    }
    
    osRecvMesg(&queue, NULL, OS_MESG_BLOCK);
    osInvalDCache(vram, size);
```
Perhaps this is me but this actually seems better so it mirrors the other usage.